src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -0,0 +1,148 @@
+import core._
+import Contexts.Context, Types._, Constants._, Decorators._, Symbols._
+import TypeUtils._, TypeErasure._
+
+
+/** Implements partial `isInstanceOf` evaluation according to the matrix on:
+ *  https://github.com/lampepfl/dotty/issues/1255
+ *
+ *  This is a generalized solution to raising an error on unreachable match
+ *  cases and warnings on other statically known results of `isInstanceOf`.
+ *
+ *  Steps taken:
+ *
+ *  1.  evalTypeApply will establish the matrix and choose the appropriate
+ *      handling for the case:
+ *  2.  a) handleStaticallyKnown
+ *      b) falseIfUnrelated with `scrutinee <:< selector`
+ *      c) handleFalseUnrelated
+ *      d) leave as is (aka `happens`)
+ *  3.  Rewrite according to step taken in `2`
+ */
+class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
+
+  import dotty.tools.dotc.ast.tpd._
+
+  def phaseName = "isInstanceOfEvaluator"
+  override def transformTypeApply(tree: TypeApply)(implicit ctx: Context, info: TransformerInfo): Tree = {
+    val defn = ctx.definitions
+
+    /** Handles the four cases of statically known `isInstanceOf`s and gives
+     *  the correct warnings, or an error if statically known to be false in
+     *  match
+     */
+    def handleStaticallyKnown(tree: Select, scrutinee: Type, selector: Type, inMatch: Boolean, pos: Position): Tree =
+      if (!(scrutinee <:< selector) && inMatch) {
+        ctx.error(
+          s"this case is unreachable due to `${selector.show}` not being a subclass of `${scrutinee.show}`",
+          Position(pos.start - 5, pos.end - 5)
+        )
+        rewrite(tree, to = false)
+      } else if (!(scrutinee <:< selector) && !inMatch) {
+        ctx.warning(
+          s"this will always yield false since `${scrutinee.show}` is not a subclass of `${selector.show}` (will be optimized away)",
+          pos
+        )
+        rewrite(tree, to = false)
+      } else if (scrutinee <:< selector && !inMatch) {
+        ctx.warning(
+          s"this will always yield true since `${scrutinee.show}` is a subclass of `${selector.show}` (will be optimized away)",
+          pos
+        )
+        rewrite(tree, to = true)
+      } else /* if (scrutinee <:< selector && inMatch) */ rewrite(tree, to = true)
+
+    /** Rewrites cases with unrelated types */
+    def handleFalseUnrelated(tree: Select, scrutinee: Type, selector: Type, inMatch: Boolean) =
+      if (inMatch) {
+        ctx.error(
+          s"will never match since `${selector.show}` is not a subclass of `${scrutinee.show}`",
+          Position(tree.pos.start - 5, tree.pos.end - 5)
+        )
+        rewrite(tree, to = false)
+      } else {
+        ctx.warning(
+          s"will always yield false since `${scrutinee.show}` is not a subclass of `${selector.show}`",
+          tree.pos
+        )
+        rewrite(tree, to = false)
+      }
+
+    /** Rewrites the select to a boolean if `to` is false or if the qualifier
```

Comment:
This logic is a bit weird `rewrite(tree, to = true)` does not mean `Rewrite tree to true` it means `Rewrite tree to true unless the type is nullable`